flink-clients/src/main/java/org/apache/flink/client/program/PerJobMiniClusterClient.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.client.program;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A Client to interact with a {@link MiniCluster} and manage its lifecycle for a single job.
+ *
+ * <p>Shuts down the MiniCluster after the JobResult has been received. It caches the JobResult to ensure the result can
+ * be retrieved even if the MiniCluster has been shut down.
+ *
+ * <p>We want to do this to:
+ * (a) ensure the MiniCluster is torn down after the job completes (to avoid resource leakage)
+ * (b) avoid race condition between the async shutdown code and the job result retrieval.
+ */
+public class PerJobMiniClusterClient extends MiniClusterClient {
```

Comment:
I think with this class we are breaking the Liskov substitution principle. A `PerJobMiniClusterClient` no longer respects the contract of the `MiniClusterClient` as we are only allowed to submit a single job. An indicator for this problem are the many check states in the methods. Hence, a `PerJobMiniClusterClient` should not be a `MiniClusterClient`. A `MiniClusterClient` could be a `PerJobMiniClusterClient`, though.

Reply:
There is now a `PerJobMiniCluster`.